Move txdb and dialects to chainlink-common/pkg/pg

reductionista opened the pull request:
[NONEVM-739](https://smartcontract-it.atlassian.net/browse/NONEVM-739)

### Requires

https://github.com/smartcontractkit/chainlink-common/pull/910

### Description

This  mostly just moves  `chainlink/core/internal/testutils/pgtest/txdb.go` and `chainlink/core/store/dialects/dialects.go`  into a new common package `chainlink-common/pkg/pg`, updating imports accordingly.
The purpose of this is so that it can be imported and used by `chainlink-solana` for unit testing Solana ORM's.

Since `txdb.go` was a bit of a mess, it has been cleaned up a bit.

- The global `init()` function invoked on package load has been replaced with `RegisterTxDb()` which accepts a dbUrl as a param instead of  automatically reading it from the `CL_DATABASE` env var. Since there are only two places in the codebase which depend on our custom `txdb` dialect of sql being registered, this can simply be called in both of those places (inside `NewSqlxDb()` and `NewConnection()`) instead of relying on the global `init()`.
- Prevents calling `sql.Register()` more than once if `RegisterTxDb()` is called more than once (calling `sql.Register()` more than once with the same driver name will result in an error).
- Replaces passing of `context.Background` to queries with passing of a new context which gets cancelled when the db is closed. (Note: db here refers to the txdb, which is just a single connection in the lower level pg db, corresponding to the dsn passed in which is a randomly generated UUID.)
- Corrected spelling errors

[NONEVM-739]: https://smartcontract-it.atlassian.net/browse/NONEVM-739?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

reductionista committed:
Move txdb.go and dialects.go to chainlink-common

Along with that:
- Replace dialects imports with pgcommon
- Replace testutils.Skip* with tests.Skip*
- Call RegisterTxDb from NewConnection
- Run goimports (somehow there were a few files on develop which were improperly formatted?)

reductionista commented:
I don't understand how this error could be happening when it tries to run the unit tests in CI:
```
Error: core/config/database_config.go:7:2: no required module provides package github.com/smartcontractkit/chainlink/v2/core/store/dialects; to add it:
	go get github.com/smartcontractkit/chainlink/v2/core/store/dialects
```
There's no such import in that file or anywhere in the repository on this branch. Somehow it must be running the CI unit tests on a branch that's different from the PR branch?? I can't seem to find a commit hash in the logs where it checks out the repository.

jmank88 commented:
> I don't understand how this error could be happening when it tries to run the unit tests in CI:
> 
> ```
> Error: core/config/database_config.go:7:2: no required module provides package github.com/smartcontractkit/chainlink/v2/core/store/dialects; to add it:
> 	go get github.com/smartcontractkit/chainlink/v2/core/store/dialects
> ```
> 
> There's no such import in that file or anywhere in the repository on this branch. Somehow it must be running the CI unit tests on a branch that's different from the PR branch?? I can't seem to find a commit hash in the logs where it checks out the repository.

Maybe try rebasing/merging, since CI runs on the merge result, not the head of the branch.

jmank88 commented on GNUmakefile:
I had no idea this existed, but it also looks confused:
1. `gofmt` is redundant given that `goimports` is already being used
2. `go mod tidy` only handles the current module
3. `gofmt`/`goimports` are meant to run when saving files

jmank88 replied:
> In addition to fixing imports, goimports also formats your code in the same style as gofmt so it can be used as a replacement for your editor's gofmt-on-save hook.

https://pkg.go.dev/golang.org/x/tools/cmd/goimports

reductionista replied:
ok, I think I should remove `gofmt` then and replace `go mod tidy` with `gomods -go mod tidy`. Sounds like the best solution for my personal setup would be to figure out how to get GoLand to run goimports with the proper `--local` argument since it doesn't seem to be doing that.  Having the `make presubmit` command seems useful in case someone doesn't have that set up properly--although admittedly it's confusing because it usually introduces a lot of extra changes to generated files which must be ignored.

jmank88 replied:
But is there a mechanism for this presubmit command to find modified files? Because we only need/want to apply it to modified files, not everything. That is normally done on save.

reductionista replied:
No, unfortunately there's not. If you think we should get rid of it, I'm okay with that I guess.

jmank88 replied:
:shrug: nbd but reducing it to just `gomods tidy` seems fine

reductionista replied:
I'll just remove it, since we already have a Makefile command that does that ( `make gomodtidy` )

reductionista committed:
Fix merge conflict with copied txdb driver

While rebasing, I found that CCIP has copied the old txdb driver from
chainlink into chainlink/integrations-tests. The problem with duplicated
code is it makes it more difficult to update. Little did they know, it
had already been updated even before they copied it. So this removes the
copy and imports the newer version from chainlink-common.

The motivation for copying rather than import appears to have come from
a lint rule that enforces not being able to import anything from an
external package with "internal" in the directory name. This still
requires that the NewSqlxDB be copied, but that's just a wrapper
around the pgcommon version which reads the db url from the chainlik
toml config, which can't be accessed in chainlink-common.

reductionista commented on integration-tests/contracts/ccipreader_test.go:
```diff
@@ -861,7 +862,7 @@ func testSetupRealContracts(
 	env changeset.DeployedEnv,
 ) ccipreaderpkg.CCIPReader {
-	db := pgtest.NewSqlxDB(t)
+	db := pgcommon.NewSqlxDB(t, pgcommon.DbUrlOrInMemory(t))
```
TODO:  add a `NewSqlxDB(t)`to pgcommon, and rename this one as a helper it calls so that we don't have to fall back to "in memory". This shouldn't ever matter since the db url should be configured automatically in integration tests, but if it did it would be better to just fail here rather than attempting to run in-memory which wouldn't work.

reductionista committed:
chainlink-common/pkg/pg -> chainlink-common/pkg/sqlutil/pg

reductionista committed:
Fix lint introduced by rebase

reductionista committed:
pg.NewSqlxDB -> pg.NewTestDB

reductionista committed:
Update chainlink-common ref

reductionista committed:
Fix merge conflict
